packages/relay-compiler/bin/RelayCompilerMain.js

```diff
@@ -174,86 +186,105 @@ Root files can include:
     throw new Error("I can't be quiet and verbose at the same time");
   }
 
-  const reporter = new ConsoleReporter({
-    verbose: options.verbose,
-    quiet: options.quiet,
+  const watchman = config.watchman && (await WatchmanClient.isAvailable());
+
+  const codegenRunner = getCodegenRunner({
+    ...config,
+    persistOutput,
+    schema,
+    src,
+    watchman,
   });
-  const useWatchman = options.watchman && (await WatchmanClient.isAvailable());
 
-  const schema = getSchema(schemaPath);
+  if (!config.validate && !config.watch && watchman) {
+    // eslint-disable-next-line no-console
+    console.log('HINT: pass --watch to keep watching for changes.');
+  }
 
-  const languagePlugin = getLanguagePlugin(options.language);
+  const result = config.watch
+    ? await codegenRunner.watchAll()
+    : await codegenRunner.compileAll();
 
-  const inputExtensions = options.extensions || languagePlugin.inputExtensions;
-  const outputExtension = languagePlugin.outputExtension;
+  if (result === 'ERROR') {
+    process.exit(100);
+  }
+  if (config.validate && result !== 'NO_CHANGES') {
+    process.exit(101);
+  }
+}
 
+function getCodegenRunner(config: Config) {
+  const reporter = new ConsoleReporter({
+    verbose: config.verbose,
+    quiet: config.quiet,
+  });
+  const schema = getSchema(config.schema);
+  const languagePlugin = getLanguagePlugin(config.language);
+  const inputExtensions = config.extensions || languagePlugin.inputExtensions;
+  const outputExtension = languagePlugin.outputExtension;
   const sourceParserName = inputExtensions.join('/');
   const sourceWriterName = outputExtension;
-
   const sourceModuleParser = RelaySourceModuleParser(
     languagePlugin.findGraphQLTags,
   );
-
-  const providedArtifactDirectory = options.artifactDirectory;
+  const providedArtifactDirectory = config.artifactDirectory;
   const artifactDirectory =
     providedArtifactDirectory != null
       ? path.resolve(process.cwd(), providedArtifactDirectory)
       : null;
-
   const generatedDirectoryName = artifactDirectory || '__generated__';
-
   const sourceSearchOptions = {
     extensions: inputExtensions,
-    include: options.include,
-    exclude: ['**/*.graphql.*', ...options.exclude], // Do not include artifacts
+    include: config.include,
+    exclude: ['**/*.graphql.*', ...config.exclude],
   };
   const graphqlSearchOptions = {
     extensions: ['graphql'],
-    include: options.include,
-    exclude: [path.relative(srcDir, schemaPath)].concat(options.exclude),
+    include: config.include,
+    exclude: [path.relative(config.src, config.schema)].concat(config.exclude),
   };
-
   const parserConfigs = {
     [sourceParserName]: {
-      baseDir: srcDir,
+      baseDir: config.src,
       getFileFilter: sourceModuleParser.getFileFilter,
       getParser: sourceModuleParser.getParser,
       getSchema: () => schema,
-      watchmanExpression: useWatchman
+      watchmanExpression: config.watchman
```

Comment:
revert? we should only pass a watchman expression  if watchman is available, not just if config.watchman is  set.